src/ui/Toolbar.tsx

```diff
@@ -10,166 +9,117 @@ type Props = {
-  };
-
-  clearSearch = () => {
-    this.setState({ search: "" });
-  };
-
-  next() {
-    let primitives = this.getPrimitives();
-    if (primitives.length == 0) return; // Nothing to select.
-    let i = this.getSelectedPrimitiveIndex(primitives);
-    // If nothing is selected, select the first primitive, otherwise select the next.
-    if (i === primitives.length - 1) i -= 1;
-    if (i === null) i = -1;
-    this.selectPrimitive(primitives[i + 1]);
-  }
-
-  prev() {
-    let primitives = this.getPrimitives();
-    if (primitives.length == 0) return; // Nothing to select.
-    let i = this.getSelectedPrimitiveIndex(primitives);
-    // If the first primitive is selected, keep it, otherwise select the previous.
-    if (i === 0) i = 1;
-    this.selectPrimitive(i === null ? null : primitives[i - 1]);
-  }
-
-  handleFocusPrimitive(selectedPrimitive: State["selectedPrimitive"]) {
-    this.setState({ selectedPrimitive: selectedPrimitive });
-  }
-
-  selectPrimitive(selectedPrimitive: null | Primitive) {
-    if (selectedPrimitive?.element) {
-      selectedPrimitive.element.focus(); // will trigger handleFocusPrimitive
+var Toolbar = (props: Props) => {
+  const { primitives, languageId, blockMode, toolbarRef } = props;
+
+  const [search, setSearch] = useState("");
+  const [selectedPrimitive, setSelectedPrimitive] = useState(null);
+  const clearSearch = () => setSearch("");
+  const changeSearch: React.ChangeEventHandler<HTMLInputElement> = (event) =>
+    setSearch(event.target.value);
+
+  // Get a flat array of all primitives matching 'search'
+  const getPrimitives = () =>
+    (primitives?.filter(search).primitives || []) as Primitive[];
+
+  // Set selectedPrimitive state, depending on whether we go up or down
+  const move = (dir: string) => {
+    let primitives = getPrimitives();
+    if (primitives.length == 0) return; // Nothing to select. Bail.
+    let i = primitives.indexOf(selectedPrimitive); // -1 if nothing selected
+    if (dir == "Down") {
+      i = Math.min(i + 1, primitives.length - 1);
     } else {
-      this.setState({ selectedPrimitive: selectedPrimitive });
+      i = Math.max(i - 1, 0);
     }
-  }
-
-  // NOTE(DS26GTE): this is just so onBlur has a non-null value
-  handleBlurPrimitive() {}
+    if (primitives[i]?.element) {
+      primitives[i].element.focus(); // should *not* apply useCallback, correct?
+    } else {
+      setSelectedPrimitive(primitives[i]);
+    }
+  };
 
-  handleKeyDown: React.KeyboardEventHandler<HTMLInputElement> = (event) => {
+  const handleKeyDown: React.KeyboardEventHandler<HTMLInputElement> = (
+    event
+  ) => {
     switch (CodeMirror.keyName(event)) {
-      case "Esc":
-        (event.target as HTMLInputElement).blur();
-        return;
       case "Down":
-        event.preventDefault();
-        this.next();
-        return;
       case "Up":
         event.preventDefault();
-        this.prev();
+        move(CodeMirror.keyName(event));
```

Comment:
If I'm going to use the computed value I'm switching over inside one of the cases, I like to extract the computed value into a standalone variable because then typescript will can statically analyze the switch statement and know exactly what the value could be when I use it in a particular case.